Closes #1357: verbatimTextOutput should optionally be hidden if no content

bborgesr opened the pull request:
Can be done with only:

* a new parameter to `verbatimTextOutput` (called `placeholder`, and defaulting to TRUE -- which means no old apps will look different from before);

* a new parameter to `textOutput` (`...` so that we can send through the new class we add in `verbatimTextOutput`);

* a little CSS.

No JS needed!

Example app (one with the default, another with `placeholder = FALSE`):

```r
library(shiny)

ui <- basicPage(
  textInput("txt", "Enter the text to display below:"),
  verbatimTextOutput("default"),
  verbatimTextOutput("new", placeholder = FALSE),
  p("Dummy text to show there is no extra spacing when", 
    "the second output is empty")
)

server <- function(input, output, session) {
  output$default <- renderText({ input$txt })
  output$new <- renderText({ input$txt })
}

shinyApp(ui, server)

```

wch commented:
Hm, it looks like there were changes to the generated JS code. I think this means you probably have a different version of Babel (or something like that) installed. Have you reinstalled the JS toolchain since we made the switch to yarn?

wch commented on R/bootstrap.R:
```diff
@@ -948,8 +953,8 @@ textOutput <- function(outputId, container = if (inline) span else div, inline =
-verbatimTextOutput <- function(outputId) {
-  textOutput(outputId, container = pre)
+verbatimTextOutput <- function(outputId, placeholder = TRUE) {
+  textOutput(outputId, container = pre, class = if (!placeholder) "noplaceholder")
```
I think it would make more sense to directly call `pre()` here instead of calling `textOutput()`.  This is so that you won't have to modify `textOuput` to accept `...`.

bborgesr replied:
Sure, that makes sense. Updated it.

wch commented:
It's nice that this can be done with just CSS!

bborgesr commented:
@wch, you were right! I hadn't updated after we switched to yarn. Everything matches up now. However, I've polluted the commit history a bit, so we should squash-merge. I've also added a NEWS entry for this feature.

On an unrelated note, the Travis check is failing in a bizarre way (and that is not reproducible locally):

```
  testthat results ================================================================
  OK: 526 SKIPPED: 0 FAILED: 1
  1. Failure: unreferenced observers are garbage collected (@test-gc.r#32) 
```

Is this Travis being weird or did I mess something up?

wch commented on R/bootstrap.R:
```diff
@@ -948,8 +951,9 @@ textOutput <- function(outputId, container = if (inline) span else div, inline =
-  textOutput(outputId, container = pre)
+verbatimTextOutput <- function(outputId, placeholder = TRUE) {
+  pre(id = outputId,
+      class = paste("shiny-text-output", if (!placeholder) "noplaceholder"))
```
A nit: if `!placeholder`, this will result in `"shiny-text-output "`, with a trailing space. One way to work around this is to use `paste()` on a vector and use the `collapse` argument, as in:

```R
paste(c("shiny-text-output", if (!placeholder) "noplaceholder"), collapse = " ")
```

wch commented:
The `placeholder` arg seems like a decent name to me. @jcheng5, what do you think?

bborgesr commented:
Used `collapse = " "` per Winston's suggestion, rebased and ready to merge!

bborgesr commented:
An better and updated example:

```r
library(shiny)

ui <- basicPage(
  textInput("txt", "Enter the text to display below:"),
  p("OLD:"),
  verbatimTextOutput("old", placeholder = TRUE, 
                     style = "background-color: #cdb38b;"),
  p("NEW (and default from now on)"),
  verbatimTextOutput("default", 
                     style = "background-color: #fa8072;"),
  p("Dummy text to show there is no extra spacing when", 
    "the second output is empty")
)

server <- function(input, output, session) {
  output$old <- renderText({ input$txt })
  output$default <- renderText({ input$txt })
}

shinyApp(ui, server)
```
